std-bits/base/src/main/java/org/enso/base/ObjectComparator.java

```diff
@@ -1,73 +1,69 @@
+      var module = Context.getCurrent().getBindings("enso").invokeMember("get_module", "Standard.Base.Data.Ordering");
+      var type = module.invokeMember("get_type", "Comparable");
 
-  public ObjectComparator() {
-    this(
-        (a) -> (b) -> {
-          throw new CompareException(a, b);
-        });
+      var are_equal = module.invokeMember("get_method", type, "compare_callback");
+      ensoCompareCallback = (v, u) -> {
+        var result = are_equal.execute(null, v, u);
+        if (result.isNull()) {
+          throw new CompareException(u, v);
+        } else {
+          return result.asInt();
+        }
+      };
+    }
   }
 
-  public ObjectComparator(Function<Object, Function<Object, Value>> fallbackComparator) {
-    this(fallbackComparator, (a) -> (b) -> Value.asValue(Text_Utils.compare_normalized(a, b)));
+  public static int ensoCompare(Object value, Object other) throws ClassCastException {
+    initCallbacks();
+    return ensoCompareCallback.apply(value, other);
   }
 
-  private ObjectComparator(Function<Object, Function<Object, Value>> fallbackComparator, Function<String, Function<String, Value>> textComparator) {
-    this.fallbackComparator = fallbackComparator;
-    this.textComparator = textComparator;
+  private final BiFunction<String, String, Integer> textComparator;
+
+  public ObjectComparator() {
+    this(true, Locale.ROOT);
   }
 
-  /**
-   * Create a copy of the ObjectComparator with case-insensitive text comparisons.
-   * @param locale to use for case folding.
-   * @return Comparator object.
-   */
-  public ObjectComparator withCaseInsensitivity(Locale locale) {
-    return new ObjectComparator(this.fallbackComparator, (a) -> (b) -> Value.asValue(Text_Utils.compare_normalized_ignoring_case(a, b, locale)));
+  public ObjectComparator(boolean caseSensitive, Locale locale) {
+    if (caseSensitive) {
+        textComparator = Text_Utils::compare_normalized;
+    } else {
+        textComparator = (a, b) -> Text_Utils.compare_normalized_ignoring_case(a, b, locale);
+    }
   }
 
-  /**
-   * Create a copy of the ObjectComparator with case-insensitive text comparisons.
-   * @param textComparator custom comparator for Text.
-   * @return Comparator object.
-   */
-  public ObjectComparator withCustomTextComparator(Function<String, Function<String, Value>> textComparator) {
-    return new ObjectComparator(this.fallbackComparator, textComparator);
+  public ObjectComparator(Function<Object, Function<Object, Value>> textComparator) {
+    this.textComparator = (a, b) -> {
+      var result = textComparator.apply(a).apply(b);
+      if (result.isNull()) {
+        throw new CompareException(a, b);
+      }
+      return result.asInt();
+    };
   }
 
   @Override
-  public int compare(Object thisValue, Object thatValue) throws ClassCastException {
+  public int compare(Object thisValue, Object thatValue) {
     // NULLs
     if (thisValue == null) {
       if (thatValue != null) {
```

Comment:
The code delegates to `EnsoObjectWrapper` and that class is using `org.graalvm.polyglot.Context`. We cannot/shouldn't use `org.graalvm.polyglot.Context` in the engine - e.g. I am afraid that isn't much to share in current implementation.

Reply:
I don't understand?

This is just a workaround that should be replaced with a proper solution. The workaround uses a Context, because it is its way to 'access' the current implementation and delegate to it (through a slow Java-to-Enso call).

The whole idea of extracting the shared code is so that the logic for computing the hash for particular primitive types can be exposed in pure Java, so that our library could call into it directly in Java, without the overhead of an Java-to-Enso call. That is what is described in #5259, although maybe I did not describe this well enough. If it's not clear - let's discuss.